From 3efc5db2a9cb60a149e213f73d6da498e4fdbfc0 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 11:59:20 -0400
Subject: [infiniband] core: Remove unneeded kobject_get/put calls

Message-id: <3daaad822b4ebc3b7e3c680eaab0f197a4049a76.1410425016.git.avadai@redhat.com>
Patchwork-id: 93677
O-Subject: [Patch RHEL7.1 V1 024/197] IB/core: Remove unneeded kobject_get/put calls
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Haggai Eran <haggaie@mellanox.com>

The ib_core module will call kobject_get on the parent object of each
kobject it creates.  This is redundant since kobject_add does that
anyway.

As a side effect, this patch should fix leaking the ports kobject and
the device kobject during unregister flow, since the previous code
didn't seem to take into account the kobject_get calls on behalf of
the child kobjects.

Signed-off-by: Haggai Eran <haggaie@mellanox.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 373c0ea181c8d277a4a51bbd705189a6a030b81c)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/core/sysfs.c b/drivers/infiniband/core/sysfs.c
index 7d3292c..1f77072 100644
--- a/drivers/infiniband/core/sysfs.c
+++ b/drivers/infiniband/core/sysfs.c
@@ -534,7 +534,7 @@ static int add_port(struct ib_device *device, int port_num,
  p->port_num   = port_num;
 
  ret = kobject_init_and_add(&p->kobj, &port_type,
-       kobject_get(device->ports_parent),
+       device->ports_parent,
        "%d", port_num);
  if (ret)
   goto err_put;
@@ -599,7 +599,6 @@ err_remove_pma:
  sysfs_remove_group(&p->kobj, &pma_group);
 
 err_put:
- kobject_put(device->ports_parent);
  kfree(p);
  return ret;
 }
@@ -835,7 +834,7 @@ int ib_device_register_sysfs(struct ib_device *device,
  }
 
  device->ports_parent = kobject_create_and_add("ports",
-     kobject_get(&class_dev->kobj));
+            &class_dev->kobj);
  if (!device->ports_parent) {
   ret = -ENOMEM;
   goto err_put;
@@ -876,8 +875,6 @@ err_put:
   }
  }
 
- kobject_put(&class_dev->kobj);
-
 err_unregister:
  device_unregister(class_dev);
 
-- 
1.7.1